Split scoring and summarising scores

sbfnk opened the pull request:
This splits `score_models` into two functions, one (called `score_forecasts`) that scores every individual forecast, and another one (called `summarise_scores`) that summarises them by model, location, target variable and horizon, and calculates relative scores.

This would
1) allow to save raw scores to the repo making it easier for teams to conduct their own investigation.
2) address the second point in #3 

As an example, the line in https://github.com/epiforecasts/covid19-forecast-hub-europe/blob/main/code/reports/rmdchunks/score-forecasts.Rmd would have to be changed to


```r
library(EuroForecastHub)
scores <- score_forecasts(
  forecasts = data,
  quantiles = get_hub_config("forecast_type")$quantiles
)
table <- summarise_scores(
  scores = scores,
  report_date = ymd(params$report_date),
  restrict_weeks = params$restrict_weeks
)
```

(or we could have a wrapper function `score_models` that maintains the original functionality and calls the new functions; or we could score the forecasts in a separate github action and read in the csv file(s) + summarise when doing the report...).

Bisaloo commented:
> we could score the forecasts in a separate github action and read in the csv file(s) + summarise when doing the report

Yes, I really like this! It makes more sense :+1:

Bisaloo commented:
I see you dropped many columns compared to our current `table` so I want to make sure I understand the workflow you have in mind.


### Option 1

1. `scores_forecasts()`
2. save output as csv
3. run `summarise_scores()` on the output of `scores_forecasts()` (either directly or by loading the csv produced elsewhere)
4. display the output directly in the evaluation reports (no more need to filter extra columns)

If this is correct, it means our csv files will no longer contain the summarised version of indices.

### Option 2

1. `scores_forecasts()`
3. run `summarise_scores()` on the output of `scores_forecasts()`
2. save output as csv
4. display the output directly in the evaluation reports (no more need to filter extra columns)

In this case, the csv files no longer contain some indices that we displayed until now only there (and not in reports), such as `overprediction` & `underprediction`. And in this case, why are we computing them in the first place?

---

In both cases, the summarised version of some indices (overprediction, underprediction, etc.) will no longer be available anywhere. Is this okay?

sbfnk commented:
My thinking was that we would go with option 1: save the "raw" scores, but not the derivative/aggregate scores (which are in the reports) which depend on the specific inclusion/exclusion criteria, which we might want to change. For example, a suggestion that came up in the last meeting was to do the aggregate scores for the report as a rolling average, i.e. only aggregate the last X weeks instead of the whole history. This could be done without having to re-run the evaluations.

sbfnk commented:
If you feel strongly re computation inside/outside the reports, we could also save the latest summaries in a separate csv and then read in that file for the reports.

Bisaloo commented:
Sorry for force-pushing, our commits crossed while I was fixing the tests on my end and integrating the latest changes from `main` as well. It resulted in a mess I could only solve this way.

Bisaloo commented:
I don't feel strongly regarding where computation is done. I just don't like so much the fact that reports save csv as a side effect. [The fact that we forgot that upload them as part as our workflow](https://github.com/epiforecasts/covid19-forecast-hub-europe/commit/040943e29dfe139a1ecb856159d6d9d54f47f9cb#diff-86b48cf4cd890334d28e917846ba4540f9b0427eac1177a349425ed42ee5e83e) is a good sign this is not a very intuitive way to proceed.

My only comment is that we will lose the aggregated version of some indices (`unprediction`, `overprediction`, etc.) If you believe this is not an issue, then it's fine by me as well.

Bisaloo reviewed (changes requested):
It looks like this change modifies the values of `rel_ae` and `rel_wis` in our evaluation report. The total number of rows differs as well (so does the total number of columns but it's expected since we only retain variables of interest in `summarise_scores()`. I'll investigate.

sbfnk commented:
> My only comment is that we will lose the aggregated version of some indices (`unprediction`, `overprediction`, etc.) If you believe this is not an issue, then it's fine by me as well.

I think it's fine to drop them for now (and bring them back if we e.g. want to plot them).

Bisaloo commented on R/summarise_scores.R:
```diff
@@ -0,0 +1,112 @@
+    group_by(model, target_variable, horizon, location) %>%
+    summarise(across(starts_with("cov_"), mean), .groups = "drop")
+  bias <- score_df %>%
+    filter(n_quantiles == 23) %>%
```
We used to compute bias even if not all quantiles were present.

https://github.com/epiforecasts/EuroForecastHub/blob/6f2dfd7dc6906c7d806ed343aed6b4731f197ae1/R/score_models.R#L92-L99

Now, we require the presence of the 23 quantiles to compute it. Was the previous approach wrong?

It seems so because it wasn't doing what it should have done according to the comment a dozen lines later:

https://github.com/epiforecasts/EuroForecastHub/blob/6f2dfd7dc6906c7d806ed343aed6b4731f197ae1/R/score_models.R#L117

sbfnk replied:
Yes, I think it was previously wrong - the values are not comparable between models if the underlying quantiles are different.

Bisaloo commented on R/summarise_scores.R:
```diff
@@ -0,0 +1,112 @@
+    cont_weeks <- score_df %>%
+      group_by(model, location, target_variable, target_end_date, horizon, forecast_date) %>%
+      summarise(present = 1, .groups = "drop") %>%
+      complete(model, location, target_variable, target_end_date, horizon) %>%
```
We used to create the complete set of possibilities with:

```r
complete(model, location, target_variable, forecast_date, horizon)
```

Upon cursory look, it seems equivalent to the current approach but it's not. `complete()` doesn't behave like `group_by()`.

I *believe* (but not 100% sure. I'm having a hard time wrapping my head around this) that `forecast_date` is the right option since we want to make sure that a forecast was posted each week, not that each week was forecasted.

Bisaloo reviewed (changes requested):
Slowly trying to understand where the differences between the new and the old versions come from. I made a couple comments on critical differences (but oh so difficult to spot...).

There is still the issue with the relative scores being different but I'd like to see if this is due to the problems mentioned here or if there's still something else.

Bisaloo commented:
Okay, after this change, the only difference in terms of lines if the fact that anomalies in truth data break the current streak of consecutive submissions (`continuous_weeks`). It's not entirely clear to me why/how it was not the case before but that's what I got from inspecting differences in the two dataframe :shrug:.

For the differences in `rel_ae` and `rel_wis`, I'm completely clueless. A possibility might be that we were computing the absolute error on point forecasts only when we are now computing it on quantiles (or it is the same for scoringutils internally?).

sbfnk commented:
> For the differences in `rel_ae` and `rel_wis`, I'm completely clueless. A possibility might be that we were computing the absolute error on point forecasts only when we are now computing it on quantiles (or it is the same for scoringutils internally?).

Yes, that makes sense regarding ae (sorry, realising now that this should really have been part of a different PR). However, it doesn't explain why `rel_wis` is different.

Bisaloo commented:
Even when setting `fc_df <- forecasts` (i.e., removing the change point point forecasts being the median), I don't get the same results so I'm clueless :shrug:. I'll take a break from this and come back with fresh eyes in a couple of days.

sbfnk commented:
Are the values of coverage and/or mean WIS (by model, if calculated from the table returned by `score_forecasts`) the same as in the current tables? That might help drilling down where the difference occurs.

sbfnk commented:
One thing worth bearing in mind is that we've replaced the calculation of the pairwise (`rel_`...) scores that's internal to `scoringutils::eval_forecasts_quantile` ([here](https://github.com/epiforecasts/scoringutils/blob/6895b49125c2dba7489c37896ac98bcacdd46e1e/R/eval_forecasts_quantile.R#L151)) with explicits call to `scoringutils::pairwise_comparison` (e.g. [here](https://github.com/epiforecasts/EuroForecastHub/blob/dbc4f9b5cc981e722cd1f5ffa91bff2e3c1e873a/R/summarise_scores.R#L75)). In doing so perhaps we have dropped something else step that's going on internally in `eval_forecast`?

Bisaloo commented:
Okay, we're slowly getting there. The current version gives the same results for `rel_ae` and `rel_wis` expected for `location == "Overall"`.

Additionally, we're using an internal function from scoringutils. This is okay-ish since we control both packages but it would be better to see if it's really not possible with `pairwise_comparison()` or ask Nikos his opinion about exporting this function.

sbfnk commented:
Prompted by your observation I just had another look at the way `scoringutils::add_rel_skill_to_eval_forecasts` operates for “Overall” and I think it’s wrong: the way we have set up the “Overall” scores I set `location` to “Overall” but keep `location_name` as its original value. What the function does in that case is to the average over all `location_name`s ([here](https://github.com/epiforecasts/scoringutils/blob/6895b49125c2dba7489c37896ac98bcacdd46e1e/R/pairwise-comparisons.R#L140)) and then apply the pairwise comparison, whereas we want the pairwise comparison to apply to the unaveraged scores. I think this is a bug in what we’ve been doing, and therefore we should revert the last commit, which also avoids calling an internal function.

What I’m not so sure yet about is if we want `summarised = FALSE` as per the previous commit (90eb5bb). Reading the documentation of the function I think we might want `summarised = TRUE` but I might be misunderstanding what that does.

Bisaloo commented:
> Prompted by your observation I just had another look at the way `scoringutils::add_rel_skill_to_eval_forecasts` operates for “Overall” and I think it’s wrong: the way we have set up the “Overall” scores I set `location` to “Overall” but keep `location_name` as its original value. What the function does in that case is to the average over all `location_name`s ([here](https://github.com/epiforecasts/scoringutils/blob/6895b49125c2dba7489c37896ac98bcacdd46e1e/R/pairwise-comparisons.R#L140)) and then apply the pairwise comparison, whereas we want the pairwise comparison to apply to the unaveraged scores. I think this is a bug in what we’ve been doing, and therefore we should revert the last commit, which also avoids calling an internal function.

I don't think this is true for 2 reasons:
- we discard the `location_name` column before calling `add_rel_skill_to_eval_forecasts()` ([here](https://github.com/epiforecasts/EuroForecastHub/pull/5/files#diff-ff775dcd488d62522aa0971ea47623b261bb210f7370b07adcf670c44ad108e7R74) and [here](https://github.com/epiforecasts/EuroForecastHub/pull/5/files#diff-ff775dcd488d62522aa0971ea47623b261bb210f7370b07adcf670c44ad108e7R89)). 
- [the averaging step in `add_rel_skill_to_eval_forecasts()` that you mention](https://github.com/epiforecasts/scoringutils/blob/6895b49125c2dba7489c37896ac98bcacdd46e1e/R/pairwise-comparisons.R#L140) only averages on columns listed in `by`, which does not include `location_name` (it doesn't even exist anymore since we discarded it just before).

> What I’m not so sure yet about is if we want `summarised = FALSE` as per the previous commit ([90eb5bb](https://github.com/epiforecasts/EuroForecastHub/commit/90eb5bb97bd5b801951e4a85e29f0f300abe5cd4)). Reading the documentation of the function I think we might want `summarised = TRUE` but I might be misunderstanding what that does.

To be honest, I'm not 100% sure about what happens here either but what I know if that in `eval_forecasts()`, `add_rel_skill_to_eval_forecasts()` (and therefore `pairwise_comparison()`) is called before the "summarising" step takes place, so on the output of `eval_forecaststs(summarise = FALSE)`.

sbfnk commented:
> I don't think this is true for 2 reasons:
> 
> * we discard the `location_name` column before calling `add_rel_skill_to_eval_forecasts()` ([here](https://github.com/epiforecasts/EuroForecastHub/pull/5/files#diff-ff775dcd488d62522aa0971ea47623b261bb210f7370b07adcf670c44ad108e7R74) and [here](https://github.com/epiforecasts/EuroForecastHub/pull/5/files#diff-ff775dcd488d62522aa0971ea47623b261bb210f7370b07adcf670c44ad108e7R89)).
> * [the averaging step in `add_rel_skill_to_eval_forecasts()` that you mention](https://github.com/epiforecasts/scoringutils/blob/6895b49125c2dba7489c37896ac98bcacdd46e1e/R/pairwise-comparisons.R#L140) only averages on columns listed in `by`, which does not include `location_name` (it doesn't even exist anymore since we discarded it just before).

I think this is precisely the problem. We end up with a table that has multiple values (representing multiple `location_name`s for each combination of `model`, `horizon` and `target_variable`) if `location=="Overall"`, which get averaged in `add_rel_skill...`. Re-reading the documentation, do we need something like `by = c("model", "target_variable", "horizon", "location", "location_name")` because we want to _summarise_ by location (i.e. "Overall" or individual countries) but individual observations/scores are at the country level and we don't want them to be averaged before doing the pairwise comparison? @nikosbosse does that sound right?
> 
> > What I’m not so sure yet about is if we want `summarised = FALSE` as per the previous commit ([90eb5bb](https://github.com/epiforecasts/EuroForecastHub/commit/90eb5bb97bd5b801951e4a85e29f0f300abe5cd4)). Reading the documentation of the function I think we might want `summarised = TRUE` but I might be misunderstanding what that does.
> 
> To be honest, I'm not 100% sure about what happens here either but what I know if that in `eval_forecasts()`, `add_rel_skill_to_eval_forecasts()` (and therefore `pairwise_comparison()`) is called before the "summarising" step takes place, so on the output of `eval_forecaststs(summarise = FALSE)`.

I just tried with `summarise = FALSE` and `summarise = TRUE` and it resulted in the same data frame.

Bisaloo commented:
I still see discrepancies I cannot explain between the old and the new version (and not just when `location == "Overall"`). I'll need to investigate.

Bisaloo commented:
Or do we expect differences in scores for all `location`s?

sbfnk commented:
In WIS or AE, or both? I think we're expecting differences in AE because we're interpreting point forecasts differently, but not in WIS.

Bisaloo commented:
In AE for some models. But you are right, this is expected.

But for WIS as well. I don't know what could cause this.

sbfnk commented:
> But for WIS as well. I don't know what could cause this.

I think the latest commit fixes this - now forecasts are also compared by forecast date. 

When I tested it, relative WIS scores seemed very similar between this and the version in `main`. Remaining differences can, I think, be explained by rounding: in the proposed version here, WIS scores are rounded before doing the pairwise comparison, whereas in `main` they are only rounded at the end.

Bisaloo reviewed (approved):
Yes, agreed, the differences are really minor. Thanks for this.

Bisaloo commented:
We need to sync this with the merge of the `split-scores` branch on `covid19-forecast-hub-europe`. Are you happy for me to take over there to try and finalize this today?

sbfnk commented:
> We need to sync this with the merge of the `split-scores` branch on `covid19-forecast-hub-europe`. Are you happy for me to take over there to try and finalize this today?

Yes, thank you - just rebased `split-scores` to main, force-pushed and PR'd: https://github.com/epiforecasts/covid19-forecast-hub-europe/pull/983
